### PR TITLE
Upgrade the source to use  ES6 syntax including classes, import and static properties

### PR DESCRIPTION
**Pull request by HosseinAgha**
Discussed at #654.
I also think this fixes the second issue in the project's backlog [here](https://github.com/Khan/KaTeX/projects/5).

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -14,7 +15,7 @@
         "indent": [2, 4, {"SwitchCase": 1}],
         "keyword-spacing": 2,
         "linebreak-style": [2, "unix"],
-        "max-len": [2, 80, 4, { "ignoreUrls": true, "ignorePattern": "\\brequire\\([\"']|eslint-disable" }],
+        "max-len": [2, 90, 4, { "ignoreUrls": true, "ignorePattern": "\\brequire\\([\"']|eslint-disable" }],
```
Comment (kevinbarabash) [inline .eslintrc]: why the change in max length?

File: .gitignore [hunk 2]
```diff
@@ -4,6 +4,7 @@ npm-debug.log
 last.png
 diff.png
 /.npm-install.stamp
+/.vscode
```
Comment (kevinbarabash) [inline .gitignore]: could add this to your global .gitignore?

**Comment by kevinbarabash** [comment]
@HosseinAgha cool!  Thanks for taking this on.

File: .eslintrc (hunk 1 above)
Comment (kevinbarabash) [inline .eslintrc]: Answering my own question: because of the extra indent.  I'm fine with this.

**Comment by kevinbarabash** [comment]
Thanks for keeping the constant definitions close to their use site.

**Comment by kevinbarabash** [comment]
@HosseinAgha after you resolve this conflict this should be good to go.

File: .eslintrc (hunk 1 above)
Comment (edemaine) [inline .eslintrc]: My two cents: I'm a fan of limiting to 80 columns.  All my terminal windows are that wide, thanks to good old punch cards...

Comment (kevinbarabash) [inline .eslintrc]: @HosseinAgha how much of the code would be affected by keeping the 80 column limit?

**Comment by HosseinAgha** [comment]
Thank you all the KaTeX team specially @kevinbarabash for all your help and patiently answering my questions.

File: .gitignore (hunk 2 above)
Comment (HosseinAgha) [inline .gitignore]: @kevinbarabash sorry I missed that one.

File: .eslintrc (hunk 1 above)
Comment (HosseinAgha) [inline .eslintrc]: @kevinbarabash Adding classes adds about 4 spaces to each line.  most of the Parser code about 13 lines have error. I think it is a lot. I decrease it to 84 but less than that makes a lot of lint errors

**Comment by HosseinAgha** [comment]
All Done

**Comment by pouriaMaleki** [comment]
How it's going? @kevinbarabash

**Comment by kevinbarabash** [comment]
I forgot about this PR :(.  I'd like to land https://github.com/Khan/KaTeX/pull/670 first b/c I think it's going to be less work to rebase this on to those changes than the other way around.  @HosseinAgha sorry about not merging this sooner and avoid extra work for you.

**Comment by kevinbarabash** [comment]
@HosseinAgha #670 has been merged.  As a result there are lots of conflicts.  :sad:  Would you be willing to rebase this?  If not, no worries.  I have some and come open a fresh PR for this.

**Comment by HosseinAgha** [comment]
@kevinbarabash Sure if we did not do anything yet.   
I had problems with server.js file you have a babel and none babel endpoint for it. Is it OK to change it to only use babel transpiled files?

**Comment by kevinbarabash** [comment]
@HosseinAgha awesome!  What kind of issues are you seeing with server.js?

**Comment by HosseinAgha** [comment]
@kevinbarabash i've described the problem at the end of [this issue](https://github.com/Khan/KaTeX/issues/654)

**Comment by kevinbarabash** [comment]
@HosseinAgha thanks or refreshing my memory.  I don't think we have much of choice here.  Could you make a separate PR for the changes to `server.js`?

**Comment by HosseinAgha** [comment]
@kevinbarabash OK

**Comment by HosseinAgha** [comment]
@kevinbarabash I created the PR (#756) for server.js fix

**Comment by HosseinAgha** [comment]
@kevinbarabash everything is merged and all tests are passing.  
Consider please consider merging this sooner because fixing merge conflicts in this scale is a bit time consuming. Thanks.

File: server.js [hunk 3]
```diff
@@ -47,8 +47,9 @@ function serveBrowserified(file, standaloneName, doBabelify) {
 }
 
 function twoBrowserified(url, file, standaloneName) {
```
Comment (kevinbarabash) [inline server.js]: Can you change the name to `browserified`?

File: server.js [hunk 4]
```diff
@@ -47,8 +47,9 @@ function serveBrowserified(file, standaloneName, doBabelify) {
...
-    app.get(url, serveBrowserified(file, standaloneName, false));
-    app.get("/babel" + url, serveBrowserified(file, standaloneName, true));
+    // we should always transpile all js files
+    // in order to support all ES2015 features
+    app.get(url, serveBrowserified(file, standaloneName, true));
 }
 
 function twoUse(url, handler) {
```
Comment (kevinbarabash) [inline server.js]: We don't need the `/babel` endpoints that this function generates either.  Can you replace all calls to `twoUse` with `app.use`?

File: .eslintrc [hunk 5]
```diff
@@ -14,7 +15,7 @@
         "indent": [2, 4, {"SwitchCase": 1}],
         "keyword-spacing": 2,
         "linebreak-style": [2, "unix"],
-        "max-len": [2, 80, 4, { "ignoreUrls": true, "ignorePattern": "\\brequire\\([\"']|eslint-disable" }],
+        "max-len": [2, 84, 4, { "ignoreUrls": true, "ignorePattern": "\\brequire\\([\"']|eslint-disable" }],
```
Comment (kevinbarabash) [inline .eslintrc]: I'm okay with this for this commit, but we should revert it afterwards.  How do other people feel?

File: src/Parser.js [hunk 6]
```diff
@@ -74,849 +53,865 @@ function ParseFuncOrArgument(result, isFunction, token) {
...
+                nucleus.text,
+                false, nucleus);
+        } else {
+            return null;
+        }
     }
-};
+}
 
 Parser.prototype.ParseNode = ParseNode;
```
Comment (kevinbarabash) [inline src/Parser.js]: Within the class you could have a line `ParseNode = ParseNode`.  This is similar to `static ParseNode = ParseNode` but adds the value to the prototype instead of the class itself.

File: src/Lexer.js [hunk 7]
```diff
@@ -84,26 +79,36 @@ const tokenRegex = new RegExp(
     ")"
 );
 
-/**
- * This function lexes a single token.
+/*
```
Comment (kevinbarabash) [inline src/Lexer.js]: super minor nit: `/**`

**Comment by kevinbarabash** [comment]
I just realized that screenshotter also uses `/babel` endpoint.  Can you update screenshotter.js to not use the `/babel` endpoint?

File: server.js (hunk 3 above)
Comment (HosseinAgha) [inline server.js]: OK as you requested I created a different PR #756 for server.js changes. could you take a look there?

**Comment by HosseinAgha** [comment]
Screenshotter changes are the also in #756

File: server.js (hunk 4 above)
Comment (HosseinAgha) [inline server.js]: As I explained in #756 isn't it better to let the codes for not transpiled versions stay there? We can use them in the future when we had a proper browser support for new ES6 features.

Comment (HosseinAgha) [inline server.js]: I think we can continue this discussion on #756

File: server.js (hunk 3 above)
Comment (HosseinAgha) [inline server.js]: Done

**Comment by HosseinAgha** [comment]
All Done

**Comment by kevinbarabash** [comment]
@HosseinAgha thanks for the PR.  I'm excited about using ES6 in our codebase.
